### PR TITLE
add batched parallel mode for IVFFLAT

### PR DESCRIPTION
**Pull request by phosphorylation**
This PR follows #2306.
In short, it adds a new parallel mode to IVFFLAT that works well for computation-intensive queries.
To use, simply set parallel_mode=4.

### CODE REVIEW

File: faiss/IndexIVFFlat.cpp [hunk 1]
```diff
@@ -166,6 +192,26 @@ struct IVFFlatScanner : InvertedListScanner {
...
+            const uint8_t* codes,
+            const idx_t* ids,
+            float* simi,
+            idx_t* idxi,
+            size_t k) const override {
+        const float* list_vecs = (const float*)codes;
+        if (metric == METRIC_INNER_PRODUCT) {
+            float_minheap_array_t res = {nx, size_t(k), idxi, simi};
+            knn_inner_product(xi, list_vecs, d, nx, list_size, &res);
+            update_res_ids<float_minheap_array_t>(&res, nx * k, ids);
```
Comment (mdouze) [inline faiss/IndexIVFFlat.cpp:206]: This should handle -1 entries. 
Please specify that the input heap should be empty (ie. not contain previous results to combine with).

File: faiss/IndexIVF.cpp [hunk 2]
```diff
@@ -625,6 +629,53 @@ void IndexIVF::search_preassigned(
...
+                    heap_heapify<HeapForL2>(k * n, distances, labels);
+                }
+            }
+#pragma omp for schedule(dynamic)
+            for (idx_t il = 0; il < invlists->nlist; il++) {
+                std::vector<idx_t>& my_list = keys_by_list[il];
+                idx_t num_of_x = my_list.size();
+                if (num_of_x > 0) {
+                    std::vector<idx_t> local_idx(k * num_of_x);
+                    std::vector<float> local_dis(k * num_of_x);
```
Comment (mdouze) [inline faiss/IndexIVF.cpp:659]: should be initialized with the neutral value (+ or - infinity)

File: faiss/IndexIVF.cpp [hunk 3]
```diff
@@ -625,6 +629,53 @@ void IndexIVF::search_preassigned(
...
+                } else {
+                    heap_heapify<HeapForL2>(k * n, distances, labels);
+                }
+            }
+#pragma omp for schedule(dynamic)
+            for (idx_t il = 0; il < invlists->nlist; il++) {
+                std::vector<idx_t>& my_list = keys_by_list[il];
+                idx_t num_of_x = my_list.size();
+                if (num_of_x > 0) {
+                    std::vector<idx_t> local_idx(k * num_of_x);
```
Comment (mdouze) [inline faiss/IndexIVF.cpp:658]: should be initilaized with -1s

**Review by mdouze — COMMENTED** [review]
Thanks! Looks legit.

Please add a test, use a
https://github.com/facebookresearch/faiss/blob/d50211a38f5f0148281da6c9e74e184c47153ed2/tests/test_index_accuracy.py#L304
for inspiration.
Please also add a benchmark to faiss/benchs that shows that this implementation is faster for some settings of ntotal / nlist / nprobe / k.

**Comment by phosphorylation** [comment]
I had problems compiling the python version, so a cpp test is provided instead. The issues with -1s are also addressed.

File: faiss/IndexIVFFlat.cpp (hunk 1 above)
Comment (phosphorylation) [reply faiss/IndexIVFFlat.cpp:209]: handled by the last commit.

File: faiss/IndexIVF.cpp (hunk 2 above)
Comment (phosphorylation) [reply faiss/IndexIVF.cpp:665]: handled by the last commit.

File: faiss/IndexIVF.cpp (hunk 3 above)
Comment (phosphorylation) [reply faiss/IndexIVF.cpp:664]: handled by the last commit.

**Comment by junjieqi** [comment]
Hi @phosphorylation what's the current status of this PR?

**Comment by asadoughi** [comment]
Closing as the PR is stale

**Pull request closed without merging**